src/vm/types/serialization.rs

```diff
@@ -0,0 +1,862 @@
+    Bool { 
+        #[serde(rename="type")]
+        type_n: String,
+        value: bool },
+    Simple { 
+        #[serde(rename="type")]
+        type_n: String,
+        value: String },
+    Container {
+        #[serde(rename="type")]
+        type_n: String,
+        value: Box<JSONParser> },
+    List {
+        #[serde(rename="type")]
+        type_n: String,
+        entries: Vec<JSONParser> },
+    Tuple {
+        #[serde(rename="type")]
+        type_n: String,
+        entries: HashMap<String, JSONParser> },
+    ContractPrincipal {
+        #[serde(rename="type")]
+        type_n: String,
+        issuer: String,
+        name: String
+    }
+}
+
+fn to_hex(val: &i128) -> String {
+    if *val >= 0 {
+        format!("{:x}", val)
+    } else {
+        format!("-{:x}", -val)
+    }
+}
+
+fn json_simple_object(type_name: &str, val: &str) -> String {
+    format!(
+        r#"{{ "type": "{}", "value": "{}" }}"#,
+        type_name, val)
+}
+
+fn json_recursive_object(type_name: &str, val: &str) -> String {
+    format!(
+        r#"{{ "type": "{}", "value": {} }}"#,
+        type_name, val)
+}
+
+impl ClaritySerializable for Value {
+    fn serialize(&self) -> String {
+        use super::Value::*;
+        use super::PrincipalData::*;
+
+        match self {
+            Int(value) => json_simple_object(TYPE_I128, &to_hex(value)),
+            Buffer(value) => json_simple_object(TYPE_BUFF, &hash::bytes_to_hex(&value.data)),
+            Bool(value) => {
+                let str_value = if *value {
+                    "true"
+                } else {
+                    "false"
+                };
+                json_recursive_object(TYPE_BOOL, str_value)
+            },
+            Principal(StandardPrincipal(data)) => {
+                json_simple_object(TYPE_STANDARD_PRINCIPAL, &data.to_address())
+            },
+            Principal(ContractPrincipal(simple_name)) => {
+                // NOTE: this should eventually panic, since "unresolved" contract principals
+                //       should never be materialized to the database.
+                format!(
+                    r#"{{ "type": "{}", "issuer": ":none:", "name": "{}" }}"#,
+                    TYPE_CONTRACT_PRINCIPAL, simple_name)
+            },
+            Principal(QualifiedContractPrincipal{ sender, name }) => {
+                format!(
+                    r#"{{ "type": "{}", "issuer": "{}", "name": "{}" }}"#,
+                    TYPE_CONTRACT_PRINCIPAL, sender.to_address(), name)
+            },
+            Response(response) => {
+                let type_name = if response.committed {
+                    TYPE_RESP_OK
+                } else {
+                    TYPE_RESP_ERR
+                };
+                let value = response.data.serialize();
+                json_recursive_object(type_name, &value)
+            },
+            Optional(OptionalData{ data: None }) => {
+                format!(r#"{{ "type": "{}" }}"#, TYPE_OPT_NONE)
```

Comment:
I ran into this problem when writing the original SIP 005 spec:  do we need to encode here what the type of the `Some` variant is?  There is ambiguity by just having `{"type": "none"}` -- `none` could be type `Option<int>` or `Option<bool>`, for example.

Reply:
Maybe this isn't important because `none` admits any type?

Reply:
For the Option Values themselves, the type doesn't need to be encoded (it's not stored in memory, for example, during evaluation) -- only lists (because they have a maximum length, separate from their actual length) and tuples (which may not actually need their type maps) store their types.